### PR TITLE
#232: improve error handling in case of exceptions during post-processing

### PR DESCRIPTION
**Pull request by mor-n4**
See commit message for more info.

See also #232.

### CODE REVIEW

**Comment by mor-n4** [comment]
This PR includes all changes of branch `GH-232-b` (in an updated version).

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSResourceValidator.java [hunk 1]
```diff
@@ -41,6 +46,37 @@
...
 	private OperationCanceledManager operationCanceledManager;
 
+	@Override
+	public List<Issue> validate(Resource resource, CheckMode mode, CancelIndicator cancelIndicator) {
+		if (resource instanceof N4JSResource) {
+			final N4JSResource resourceCasted = (N4JSResource) resource;
+			try {
+				resourceCasted.performPostProcessing(cancelIndicator);
+			} catch (Throwable th) {
+				// ignore this exception/error (we will create an issue for it below)
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSResourceValidator.java]: You could also store the exception in the `catch` via a local variable here and do not need an extra field for `postProcessingThrowable ` in `PostProcessingAwareResource.java`

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSResourceValidator.java [hunk 2]
```diff
@@ -41,6 +46,37 @@
...
+					&& resourceCasted.getPostProcessingThrowable() != null) {
+				// When getting here, we have an attempt to validate a resource that was post-processed but the
+				// post-processing failed (i.e. postProcessingThrowable!=null). Validating such a resource will usually
+				// cause a multitude of follow-up exceptions in many @Check methods (could easily be hundreds). Since,
+				// EMF/Xtext has the behavior of not aborting the overall validation due to an exception in a single
+				// @Check method, this will lead to many exceptions, which are all just follow-up issues of the problem
+				// that caused the post-processing to fail.
+				// 1) Since this is annoying and misleading, we just ignore all ordinary validation.
+				// 2) To not overlook the problem, we create a single validation issue (error) pointing out the problem.
+				// (for a test, see class AvoidFollowUpExceptionsInValidationTest)
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSResourceValidator.java]: link to `AvoidFollowUpExceptionsInValidationTest `?

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSResourceValidator.java [hunk 3]
```diff
@@ -63,4 +99,24 @@ protected void validate(Resource resource, CheckMode mode, CancelIndicator cance
...
+		final String trace = "\n" + Stream.of(th.getStackTrace())
+				.map(ste -> ste.toString())
+				.collect(Collectors.joining("\n")); // cannot add indentation, because Xtext would reformat the message
+		final String msg = IssueCodes.getMessageForPOST_PROCESSING_FAILED(thKind, thName, th.getMessage() + trace);
+		final Issue.IssueImpl issue = new Issue.IssueImpl();
+		issue.setCode(IssueCodes.POST_PROCESSING_FAILED);
+		issue.setSeverity(IssueCodes.getDefaultSeverity(IssueCodes.POST_PROCESSING_FAILED));
+		issue.setMessage(msg);
+		issue.setUriToProblem(EcoreUtil.getURI(res.getScript()));
+		issue.setType(CheckType.FAST);
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSResourceValidator.java]: A comment for CheckType.FAST ???

File: tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/validation/AvoidFollowUpExceptionsInValidationTest.xtend [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import org.junit.Test
+
+/**
+ * This tests the early return in method {@link N4JSResourceValidator#validate(Resource, CheckMode, CancelIndicator)}.
+ */
+@InjectWith(N4JSInjectorProviderWithPostProcessingFailure) // NOTE the special injector provider!
+class AvoidFollowUpExceptionsInValidationTest extends AbstractN4JSTest {
+
+	@Test
+	def void testEditorUpdate() throws Exception {
```
Comment (qtran-n4) [inline tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/validation/AvoidFollowUpExceptionsInValidationTest.xtend]: Better "testSimulateEditorUpdate"?

File: tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/validation/AvoidFollowUpExceptionsInValidationTest.xtend [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+		// Xtext always triggers a validation
+		// see last finally block in method: XtextDocument.XtextDocumentLocker#modify(IUnitOfWork<T, XtextResource>)
+		// which always invokes: XtextDocument#checkAndUpdateAnnotations()
+		val IResourceValidator validator = (res as XtextResource).getResourceServiceProvider()
+				.getResourceValidator();
+		val issues = validator.validate(res, CheckMode.ALL, CancelIndicator.NullImpl); // <-- ASSERTION #1: this should not throw follow-up exceptions
+
+		// ASSERTION #2: the exception thrown during post-processing should show up as a single error
+		assertEquals("incorrect number of issues", 1, issues.size);
```
Comment (qtran-n4) [inline tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/validation/AvoidFollowUpExceptionsInValidationTest.xtend]: Nice test!

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSResourceValidator.java (hunk 1 above)
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSResourceValidator.java]: No, this isn't possible. In the standard case, post-processing will already have taken place when method `#validate()` is entered.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSResourceValidator.java (hunk 2 above)
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSResourceValidator.java]: in plain comments, links are not possible, afaik

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSResourceValidator.java (hunk 3 above)
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSResourceValidator.java]: done

File: tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/validation/AvoidFollowUpExceptionsInValidationTest.xtend (hunk 4 above)
Comment (mor-n4) [reply tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/validation/AvoidFollowUpExceptionsInValidationTest.xtend]: done

**Pull request merged**